torchgeo/datasets/rioxarray.py

```diff
@@ -0,0 +1,156 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""In-memory geographical xarray.DataArray."""
+
+import glob
+import os
+import re
+import sys
+from datetime import datetime
+from typing import Any, Callable, Optional, cast
+
+import netCDF4  # noqa: F401
+import numpy as np
+import torch
+import xarray as xr
+from rasterio.crs import CRS
+from rtree.index import Index, Property
+
+from .geo import GeoDataset
+from .utils import BoundingBox
+
+
+class RioXarrayDataset(GeoDataset):
+    """Wrapper for geographical datasets stored as Xarray Datasets.
+
+    Relies on rioxarray.
+    """
+
+    filename_glob = "*"
+    filename_regex = ".*"
+
+    def __init__(
+        self,
+        root: str,
+        data_variables: list[str],
+        crs: Optional[CRS] = None,
+        res: Optional[float] = None,
+        transforms: Optional[Callable[[dict[str, Any]], dict[str, Any]]] = None,
+    ) -> None:
+        """Initialize a new Dataset instance.
+
+        Args:
+            root: directory with nc files
+            data_variables: data variables that should be gathered from the xr_datasets
+            crs: :term:`coordinate reference system (CRS)` to warp to
+                (defaults to the CRS of dataarray)
+            res: resolution of the dataset in units of CRS
+                (defaults to the resolution of the dataarray)
+            transforms: a function/transform that takes an input sample
+                and returns a transformed version
+
+        Raises:
+            FileNotFoundError: if no files are found in ``root``
+        """
+        super().__init__(transforms)
+
+        self.root = root
+        self.data_variables = data_variables
+        self.transforms = transforms
+
+        # Create an R-tree to index the dataset
+        self.index = Index(interleaved=False, properties=Property(dimension=3))
+
+        # Populate the dataset index
+        i = 0
+        pathname = os.path.join(root, self.filename_glob)
+        filename_regex = re.compile(self.filename_regex, re.VERBOSE)
+        for filepath in glob.iglob(pathname, recursive=True):
+            match = re.match(filename_regex, os.path.basename(filepath))
+            if match is not None:
+                with xr.open_dataset(filepath, decode_times=True) as ds:
+                    if crs is None:
+                        crs = ds.rio.crs
+                    if res is None:
+                        res = ds.rio.resolution()[0]
+
+                    (minx, miny, maxx, maxy) = ds.rio.bounds()
+
+                if hasattr(ds, "time"):
+                    try:
+                        indices = ds.indexes["time"].to_datetimeindex()
+                    except AttributeError:
+                        indices = ds.indexes["time"]
```

Comment:
Is this a standard attribute name or just something chosen by the authors of your particular NetCDF files? If necessary we can always add a class attribute that lists the names of layers to look for things like time from.

Reply:
I have encountered different naming schemes for spatial coordinate `x and y` or `lat and lon` but for the time dimension it was always called `time`. However, the `ds.rio.clip_box()` function in the `__getitem__` method expects the spatial coordinates to be named x and y otherwise it complains and tells you to rename them to x and y.